bpf: fix skip_tunnel_nodeport_revnat

lmb opened the pull request:
bpf: fix skip_tunnel_nodeport_revnat

    The test contains a bug since it passes min_port and max_port in struct
    ip{4,6}_nat_target in network instead of host byte order. It doesn't fail
    because of a series of "fortunate" circumstances, which boil down to
    invoking the following function

        __u16 __snat_try_keep_port(__u16 start, __u16 end, __u16 val)
       {
           return val >= start && val <= end ? val :
                __snat_clamp_port_range(start, end, (__u16)get_prandom_u32());
       }

    with the following arguments:

        __snat_try_keep_port(htons(port), htons(port), port)

    since htons(port) <= port <= htons(port) is (almost) never true, we end up
    invoking

        __u16 __snat_clamp_port_range(__u16 start, __u16 end,  __u16 val)
       {
           return (val % (__u16)(end - start)) + start;
       }

    Since start == end, we end up executing val % 0. This seems to cause UB
    which somehow leads to the code just using val. Of course, this all comes
    crashing down when trying to modify any of the code in this path. Fix the
    test by converting to host byte order as appropriate.

    Fixes: c1761f75ec ("bpf: Implement handling of flag_skip_tunnel for revNAT
    nodeport traffic") Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

gentoo-root commented on bpf/tests/skip_tunnel_nodeport_revnat.c:
> There is a subtle off-by-one error here. __snat_try_keep_port seems to
accept the range [start, end] while __snat_clamp_port_range is only
defined for [start, end).

Why don't you fix __snat_clamp_port_range as well to divide by `end-start+1`?

> Since val is just port, __snat_try_keep_port returns the value which
the test expects.

This part is not clear to me. __snat_clamp_port_range will return `val+start`, that is, `port+(__u16)get_prandom_u32()`. Why is this the expected number if it's random?

lmb replied:
> Why don't you fix __snat_clamp_port_range as well to divide by end-start+1?

I have a follow up PR which also fixes the problem.

> This part is not clear to me.

Hmm good point, something doesn't add up. It actually returns `prandom() + htons(port)`. Disassembly from a recent-ish commit on main: https://gist.github.com/lmb/b3ae7b8c14d4c1a150086a69161a3d49

All of this https://github.com/cilium/cilium/blob/522564b37bc6296db8be7519189ca369bcd391b3/bpf/lib/nat.h#L167-L202 is compiled to the following:

```
	w8 = 41090
; 	rstate.to_dport = otuple->sport;
	*(u16 *)(r10 - 44) = r8
; 	       __snat_clamp_port_range(start, end, (__u16)get_prandom_u32());
	call 7
	r1 = 1
; 	ostate->common.needs_ct = needs_ct;
	*(u64 *)(r10 - 144) = r1
; 	rstate.common.needs_ct = needs_ct;
	*(u64 *)(r10 - 72) = r1
; 	rstate.common.created = bpf_mono_now();
	call 5
; 		rtuple.dport = bpf_htons(port);
	*(u16 *)(r10 - 88) = r8
; 	rstate.common.created = bpf_mono_now();
	r0 /= 1000000000
	*(u64 *)(r10 - 80) = r0
	r2 = r10
	r2 += -96
	r3 = r10
	r3 += -80
; 	return map_update_elem(map, tuple, state, BPF_NOEXIST);
	r1 = 0 ll
	w4 = 1
	call 2
; 		if (__snat_create(map, &rtuple, &rstate) == 0)
	if w0 == 0 goto +1145 <LBB27_128>
```

So somehow clang just decides to completely ditch `__snat_clamp_port_range` for the first port it tries (which is the one the test expects). The only remnant seems to be the call to get_prandom_u32 (`call 7`) but the result of that call is never even used. I'm a bit baffled.

gentoo-root replied:
> I have a follow up PR which also fixes the problem.

Cool, please add me to it when it's ready.

> So somehow clang just decides to completely ditch __snat_clamp_port_range for the first port it tries

What's clang's position on modulo zero? Maybe it considers it undefined behavior and mangles the whole expression? I mean, even if it's well-defined in BPF, it can still be a UB in C, allowing the compiler to do weird stuff. It's just a hypothesis.

lmb replied:
Hmm yeah seems pretty UB to me: https://godbolt.org/z/9KbsT7j78 I'll update the commit message appropriately.

gentoo-root commented on bpf/tests/skip_tunnel_nodeport_revnat.c:
```diff
@@ -192,8 +192,8 @@ setup(struct __ctx_buff *ctx, bool v4, bool flag_skip_tunnel)
 			.addr = NODEPORT_IPV4,
-			.min_port = NODEPORT_NAT_PORT,
-			.max_port = NODEPORT_NAT_PORT,
+			.min_port = __bpf_ntohs(NODEPORT_NAT_PORT),
```
Nice catch! I think we need to introduce some naming convention for constants and variables in network byte order, so that there is smaller chance of mixing them up.

lmb committed:
bpf: fix skip_tunnel_nodeport_revnat

The test contains a bug since it passes min_port and max_port in
struct ip{4,6}_nat_target in network instead of host byte order.
It doesn't fail because of a series of "fortunate" circumstances,
which boil down to invoking the following function

    __u16 __snat_try_keep_port(__u16 start, __u16 end, __u16 val)
    {
        return val >= start && val <= end ? val :
             __snat_clamp_port_range(start, end, (__u16)get_prandom_u32());
    }

with the following arguments:

    __snat_try_keep_port(htons(port), htons(port), port)

since htons(port) <= port <= htons(port) is (almost) never true, we
end up invoking

    __u16 __snat_clamp_port_range(__u16 start, __u16 end,  __u16 val)
    {
        return (val % (__u16)(end - start)) + start;
    }

Since start == end, we end up executing val % 0. This seems to cause
UB which somehow leads to the code just using val. Of course, this all
comes crashing down when trying to modify any of the code in this path.
Fix the test by converting to host byte order as appropriate and by
extending the valid port range by one.

Fixes: c1761f75ec ("bpf: Implement handling of flag_skip_tunnel for revNAT nodeport traffic")
Signed-off-by: Lorenz Bauer <lmb@isovalent.com>
